File: tool/multiedit.go
package tool

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

// MultiEditTool creates the multi-edit tool for making multiple edits to a single file
func MultiEditTool() *ToolDefinition {
	return &ToolDefinition{
		ID:   "multiedit",
		Name: "multiedit",
		Description: `This is a tool for making multiple edits to a single file in one operation. It allows you to perform multiple find-and-replace operations efficiently. Prefer this tool over the Edit tool when you need to make multiple edits to the same file.

Before using this tool:

1. Use the Read tool to understand the file's contents and context
2. Verify the directory path is correct

To make multiple file edits, provide the following:
1. file_path: The absolute path to the file to modify (must be absolute, not relative)
2. edits: An array of edit operations to perform, where each edit contains:
   - old_string: The text to replace (must match the file contents exactly, including all whitespace and indentation)
   - new_string: The edited text to replace the old_string
   - replace_all: Replace all occurrences of old_string. This parameter is optional and defaults to false.

IMPORTANT:
- All edits are applied in sequence, in the order they are provided
- Each edit operates on the result of the previous edit
- All edits must be valid for the operation to succeed - if any edit fails, none will be applied
- This tool is ideal when you need to make several changes to different parts of the same file

CRITICAL REQUIREMENTS:
1. All edits follow the same requirements as the single Edit tool
2. The edits are atomic - either all succeed or none are applied
3. Plan your edits carefully to avoid conflicts between sequential operations

WARNING:
- The tool will fail if old_string doesn't match the file contents exactly (including whitespace)
- The tool will fail if old_string and new_string are the same
- Since edits are applied in sequence, ensure that earlier edits don't affect the text that later edits are trying to find

When making edits:
- Ensure all edits result in idiomatic, correct code
- Do not leave the code in a broken state
- Always use absolute file paths (starting with /)
- Only use emojis if the user explicitly requests it. Avoid adding emojis to files unless asked.
- Use replace_all for replacing and renaming strings across the file. This parameter is useful if you want to rename a variable for instance.`,
		InputSchema: map[string]interface{}{
			"type": "object",
			"properties": map[string]interface{}{
				"file_path": map[string]interface{}{
					"type":        "string",
					"description": "The absolute path to the file to modify",
				},
				"edits": map[string]interface{}{
					"type":        "array",
					"description": "Array of edit operations to perform sequentially on the file",
					"items": map[string]interface{}{
						"type": "object",
						"properties": map[string]interface{}{
							"old_string": map[string]interface{}{
								"type":        "string",
								"description": "The text to replace",
							},
							"new_string": map[string]interface{}{
								"type":        "string",
								"description": "The text to replace it with (must be different from old_string)",
							},
							"replace_all": map[string]interface{}{
								"type":        "boolean",
								"description": "Replace all occurrences of old_string (default false)",
							},
						},
						"required": []string{"old_string", "new_string"},
					},
				},
			},
			"required": []string{"file_path", "edits"},
		},
		Execute: executeMultiEdit,
	}
}

// EditOperation represents a single edit operation
type EditOperation struct {
	OldString  string
	NewString  string
	ReplaceAll bool
}

func executeMultiEdit(params map[string]interface{}, ctx ToolContext) (ToolResult, error) {
	filePath, ok := params["file_path"].(string)
	if !ok {
		return ToolResult{}, fmt.Errorf("file_path parameter is required")
	}

	// Make path absolute if it isn't
	if !filepath.IsAbs(filePath) {
		cwd, err := os.Getwd()
		if err != nil {
			return ToolResult{}, fmt.Errorf("failed to get current directory: %v", err)
		}
		filePath = filepath.Join(cwd, filePath)
	}

	// Parse edits array
	editsParam, ok := params["edits"].([]interface{})
	if !ok {
		return ToolResult{}, fmt.Errorf("edits parameter is required and must be an array")
	}

	if len(editsParam) == 0 {
		return ToolResult{}, fmt.Errorf("edits array cannot be empty")
	}

	// Convert to EditOperation structs
	var edits []EditOperation
	for i, editInterface := range editsParam {
		editMap, ok := editInterface.(map[string]interface{})
		if !ok {
			return ToolResult{}, fmt.Errorf("edit at index %d is not a valid object", i)
		}

		oldString, ok := editMap["old_string"].(string)
		if !ok {
			return ToolResult{}, fmt.Errorf("edit at index %d is missing old_string", i)
		}

		newString, ok := editMap["new_string"].(string)
		if !ok {
			return ToolResult{}, fmt.Errorf("edit at index %d is missing new_string", i)
		}

		if oldString == newString {
			return ToolResult{}, fmt.Errorf("edit at index %d has identical old_string and new_string", i)
		}

		replaceAll := false
		if replaceAllParam, ok := editMap["replace_all"].(bool); ok {
			replaceAll = replaceAllParam
		}

		edits = append(edits, EditOperation{
			OldString:  oldString,
			NewString:  newString,
			ReplaceAll: replaceAll,
		})
	}

	// Read the file
	content, err := os.ReadFile(filePath)
	if err != nil {
		if os.IsNotExist(err) {
			// If file doesn't exist and first edit has empty old_string, create new file
			if edits[0].OldString == "" {
				content = []byte{}
			} else {
				return ToolResult{}, fmt.Errorf("file not found: %s", filePath)
			}
		} else {
			return ToolResult{}, fmt.Errorf("failed to read file: %v", err)
		}
	}

	// Check if path is a directory
	info, err := os.Stat(filePath)
	if err == nil && info.IsDir() {
		return ToolResult{}, fmt.Errorf("path is a directory, not a file: %s", filePath)
	}

	// Apply edits sequentially
	currentContent := string(content)
	var editResults []string

	for i, edit := range edits {
		newContent, err := applyEdit(currentContent, edit)
		if err != nil {
			return ToolResult{}, fmt.Errorf("edit %d failed: %v", i+1, err)
		}
		currentContent = newContent
		editResults = append(editResults, fmt.Sprintf("Edit %d: Applied successfully", i+1))
	}

	// Write the modified content back to the file
	err = os.WriteFile(filePath, []byte(currentContent), 0644)
	if err != nil {
		return ToolResult{}, fmt.Errorf("failed to write file: %v", err)
	}

	// Get relative path for title if possible
	cwd, _ := os.Getwd()
	relPath, err := filepath.Rel(cwd, filePath)
	if err != nil {
		relPath = filePath
	}

	output := fmt.Sprintf("Successfully applied %d edit(s) to %s\n\n%s",
		len(edits), relPath, strings.Join(editResults, "\n"))

	return ToolResult{
		Title:  relPath,
		Output: output,
		Metadata: map[string]interface{}{
			"file_path":   filePath,
			"edits_count": len(edits),
		},
	}, nil
}

// applyEdit applies a single edit operation to the content
func applyEdit(content string, edit EditOperation) (string, error) {
	// Handle file creation case (empty old_string)
	if edit.OldString == "" {
		if content != "" {
			return "", fmt.Errorf("old_string is empty but file already has content")
		}
		return edit.NewString, nil
	}

	// Check if old_string exists in content
	if !strings.Contains(content, edit.OldString) {
		return "", fmt.Errorf("old_string not found in content")
	}

	// Count occurrences
	count := strings.Count(content, edit.OldString)

	if edit.ReplaceAll {
		// Replace all occurrences
		return strings.ReplaceAll(content, edit.OldString, edit.NewString), nil
	}

	// For single replacement, ensure uniqueness
	if count > 1 {
		return "", fmt.Errorf("old_string found %d times, use replace_all: true or provide more context to make it unique", count)
	}

	// Replace single occurrence
	return strings.Replace(content, edit.OldString, edit.NewString, 1), nil
}

File: ai/tools/multiedit.ts
/**
 * MultiEdit tool for performing multiple string replacements atomically.
 *
 * Allows multiple find-replace operations on a single file in one atomic operation.
 * Each edit is applied in sequence, with subsequent edits operating on the result
 * of previous edits.
 */

import { tool } from '../../node_modules/ai/dist/index.mjs';
import { z } from 'zod';
import { resolveAndValidatePathSecure, getRelativePath, fileExists } from './filesystem';
import { getFileTracker, updateFileTracker } from '../../core/state';

// Error messages
const ERROR_FILE_PATH_REQUIRED = 'file_path parameter is required';
const _ERROR_EDITS_REQUIRED = 'edits parameter is required and must be an array';
const ERROR_EDITS_EMPTY = 'edits array cannot be empty';
const ERROR_EDIT_INVALID = 'edit at index {} is not a valid object';
const ERROR_EDIT_MISSING_OLD = 'edit at index {} is missing old_string';
const ERROR_EDIT_MISSING_NEW = 'edit at index {} is missing new_string';
const ERROR_EDIT_SAME_OLD_NEW = 'edit at index {} has identical old_string and new_string';
const ERROR_EDIT_FAILED = 'edit {} failed: {}';
const ERROR_OLD_STRING_NOT_FOUND = 'old_string not found in file content';
const ERROR_OLD_STRING_MULTIPLE = 'old_string found multiple times - use replace_all or provide more context';

interface EditOperation {
  oldString: string;
  newString: string;
  replaceAll: boolean;
}

interface MultiEditResult {
  success: boolean;
  filePath: string;
  editCount: number;
  error?: string;
  output?: string;
}

function validateEdits(edits: Array<{ oldString: string; newString: string; replaceAll?: boolean }>): [EditOperation[], string | null] {
  if (!edits || edits.length === 0) {
    return [[], ERROR_EDITS_EMPTY];
  }

  const operations: EditOperation[] = [];

  for (let i = 0; i < edits.length; i++) {
    const edit = edits[i];

    if (!edit || typeof edit !== 'object') {
      return [[], ERROR_EDIT_INVALID.replace('{}', String(i))];
    }

    if (edit.oldString === undefined || edit.oldString === null) {
      return [[], ERROR_EDIT_MISSING_OLD.replace('{}', String(i))];
    }

    if (edit.newString === undefined || edit.newString === null) {
      return [[], ERROR_EDIT_MISSING_NEW.replace('{}', String(i))];
    }

    if (edit.oldString === edit.newString) {
      return [[], ERROR_EDIT_SAME_OLD_NEW.replace('{}', String(i))];
    }

    operations.push({
      oldString: edit.oldString,
      newString: edit.newString,
      replaceAll: edit.replaceAll ?? false,
    });
  }

  return [operations, null];
}

async function multieditImpl(
  filePath: string,
  edits: Array<{ oldString: string; newString: string; replaceAll?: boolean }>,
  workingDir?: string,
  sessionId?: string
): Promise<MultiEditResult> {
  // Validate file path
  if (!filePath) {
    return {
      success: false,
      filePath: '',
      editCount: 0,
      error: ERROR_FILE_PATH_REQUIRED,
    };
  }

  // Validate and resolve path (secure version that follows symlinks)
  const [absPath, pathError] = await resolveAndValidatePathSecure(filePath, workingDir);
  if (pathError) {
    return {
      success: false,
      filePath,
      editCount: 0,
      error: pathError,
    };
  }

  // Validate edits
  const [operations, validationError] = validateEdits(edits);
  if (validationError) {
    return {
      success: false,
      filePath,
      editCount: 0,
      error: validationError,
    };
  }

  const relPath = getRelativePath(absPath, workingDir);

  // Check read-before-write for existing files
  if (sessionId && await fileExists(absPath)) {
    try {
      const tracker = await getFileTracker(sessionId);
      const lastRead = tracker.readTimes.get(absPath);
      if (!lastRead) {
        return {
          success: false,
          filePath: relPath,
          editCount: 0,
          error: 'File has not been read in this session. Read the file first.',
        };
      }
    } catch (error) {
      // If session doesn't exist, skip the check
      console.warn('Failed to get file tracker:', error);
    }
  }

  try {
    // Read current content (or start with empty for new files)
    let content = '';
    const exists = await fileExists(absPath);

    if (exists) {
      const file = Bun.file(absPath);
      content = await file.text();
    }

    // Apply edits sequentially
    for (let i = 0; i < operations.length; i++) {
      const op = operations[i]!;

      // Handle file creation (empty oldString on first edit)
      if (op.oldString === '' && !exists && i === 0) {
        content = op.newString;
        continue;
      }

      // Reject empty oldString on existing files
      if (op.oldString === '' && exists) {
        // Write any successful edits before returning error
        if (i > 0) {
          await Bun.write(absPath, content);
        }
        return {
          success: false,
          filePath: relPath,
          editCount: i,
          error: ERROR_EDIT_FAILED.replace('{}', String(i + 1)).replace('{}', 'empty old_string not allowed on existing files'),
        };
      }

      // Check if oldString exists
      const occurrences = content.split(op.oldString).length - 1;

      if (occurrences === 0) {
        // Write any successful edits before returning error
        if (i > 0) {
          await Bun.write(absPath, content);
        }
        return {
          success: false,
          filePath: relPath,
          editCount: i,
          error: ERROR_EDIT_FAILED.replace('{}', String(i + 1)).replace('{}', ERROR_OLD_STRING_NOT_FOUND),
        };
      }

      // Apply the edit
      if (op.replaceAll) {
        content = content.split(op.oldString).join(op.newString);
      } else {
        // Always replace first occurrence, don't error on multiple
        content = content.replace(op.oldString, op.newString);
      }
    }

    // Write the result
    await Bun.write(absPath, content);

    // Update tracker
    if (sessionId) {
      try {
        const file = Bun.file(absPath);
        const stats = await file.stat();
        await updateFileTracker(sessionId, absPath, Date.now(), stats.mtime.getTime());
      } catch (error) {
        // Ignore tracker update errors - don't fail the operation
        console.warn('Failed to update file tracker:', error);
      }
    }

    return {
      success: true,
      filePath: relPath,
      editCount: operations.length,
      output: `Applied ${operations.length} edit(s) to ${relPath}`,
    };
  } catch (error) {
    return {
      success: false,
      filePath: relPath,
      editCount: 0,
      error: `Failed to apply edits: ${error}`,
    };
  }
}

const multieditParameters = z.object({
  filePath: z.string().describe('Absolute path to the file to modify'),
  edits: z.array(z.object({
    oldString: z.string().describe('Text to replace (must match exactly)'),
    newString: z.string().describe('Replacement text'),
    replaceAll: z.boolean().optional().describe('Replace all occurrences (default: false)'),
  })).describe('Array of edit operations to apply sequentially'),
});

export const multieditTool = tool({
  description: `Perform multiple find-and-replace operations on a single file atomically.

All edits are validated before any are applied. Each edit operates on the result
of the previous edit, allowing dependent changes.

IMPORTANT:
- Read the file first before using this tool
- old_string must match exactly (including whitespace)
- Edits are applied in sequence - plan carefully to avoid conflicts

To create a new file: use empty old_string with the file contents as new_string.`,
  parameters: multieditParameters,
  // @ts-expect-error - Zod v4 type inference issue with AI SDK
  execute: async (args: z.infer<typeof multieditParameters>) => {
    const result = await multieditImpl(args.filePath, args.edits);
    return result.success
      ? result.output!
      : `Error: ${result.error}`;
  },
});

/**
 * Create a multiedit tool with context (sessionId and workingDir) bound to it.
 */
export function createMultieditTool(context: { sessionId: string; workingDir: string }) {
  return tool({
    description: multieditTool.description,
    parameters: multieditParameters,
    // @ts-expect-error - Zod v4 type inference issue with AI SDK
    execute: async (args: z.infer<typeof multieditParameters>) => {
      const result = await multieditImpl(
        args.filePath,
        args.edits,
        context.workingDir,
        context.sessionId
      );
      return result.success
        ? result.output!
        : `Error: ${result.error}`;
    },
  });
}

export { multieditImpl };
